internal/internal_workflow.go

```diff
@@ -1123,8 +1134,10 @@ func (d *dispatcherImpl) ExecuteUntilAllBlocked(deadlockDetectionTimeout time.Du
 			}
 		}
```

Comment:
In Typescript, the Update handler preempts _immediately_ the code that was executing the  `setHandler` call, whereas here we're going to let all outstanding coroutines run until they're blocked/exit, and only then turn to the queued eager Update handler. I think that's going to result in (subtly) different semantics, e.g. related to whether or not an Update handler will see changes made by other concurrent routines. I'm not necessarily suggesting a change; that may fall well within expected inter-SDK variability. But, it would be technically possible I think to execute the pending eager coroutines inside the inner for loop, rather than waiting for that loop to exit (and I suspect it would bring Typescript and Go semantics closer together, but I haven't done the work to come up with an example.)

Reply:
Yes no coroutine in the Go SDK executes immediately when created. The same is true if users create  there own coroutine or resolve a future. The runtimes schedule work differently. I agree it is probably technically possible to make the updates more  "eager", but it isn't required to get the behavior WRT updates running before the main workflow function and allowing updates to run after being registered before the main function is finished.

Reply:
> It isn't required to get the behavior WRT updates running before the main workflow function and allowing updates to run after being registered before the main function is finished.

Hm, I think that's not true. Our requirements for signal and update hold that pending signals and updates must get an opportunity to influence the workflow return value, and the params passed to CAN. But the workflow return value may be computed by one of the pending coroutines in the inner loop. Similarly, one of the pending coroutines in the inner loop could be a signal or update handler that CANs.

I'm thinking that we can meet that requirement by ensuring that, when we yield back to the executor loop in `setUpdateHandler`, the new handler coroutine is the very next coroutine to be handled.

Here's a failing test that demonstrates what I'm saying. I think that because your test workflows are using `workflow.Await` before the end, they are passing, but that we actually have a stronger requirement.

```go
func (w *Workflows) UpdateSetHandlerOnly(ctx workflow.Context) (int, error) {
	updatesRan := 0
	updateHandle := func(ctx workflow.Context) error {
		updatesRan++
		return nil
	}
	workflow.SetUpdateHandler(ctx, "update", updateHandle)
	return updatesRan, nil
}
func (ts *IntegrationTestSuite) TestUpdateAlwaysHandled() {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	options := ts.startWorkflowOptions("test-update-always-handled")
	options.StartDelay = time.Hour
	run, err := ts.client.ExecuteWorkflow(ctx, options, ts.workflows.UpdateSetHandlerOnly)
	ts.NoError(err)
	// Send an update before the first workflow task
	_, err = ts.client.UpdateWorkflow(ctx, run.GetID(), run.GetRunID(), "update")
	ts.NoError(err)
	// ts.NoError(updateHandle.Get(ctx, nil))
	var result int
	ts.NoError(run.Get(ctx, &result))
	ts.Equal(0, result)
}
```

The equivalent test in Typescript passes: https://github.com/dandavison/temporalio-sdk-typescript/blob/125241bbd3a1b3cf29e59d47eaa0e979ad3d7e69/packages/test/src/test-integration-update.ts#L217-L243

Reply:
This test passes with my changes

Reply:
No worries, I don't see how CAN is different then returning any normal value from the workflow function